### ISSUE TITLE
What is a stream anyway?

### ISSUE THREAD

**Issue by bmcfee** [bmcfee/pescador#7]
Currently, a data stream is expected to generate samples of the forms:

- `x[i]`
- `x[i], y[i]`

where `x[i]` corresponds to a single example feature vector, and `y[i]` is the label for a single point.

This ultimately makes composing streams a little clumsy, since the shape of the data can be ambiguous.  Imagine the case where the problem is unsupervised, and `x[i].ndim == 2` (eg, clustering image patches).    There's no way to tell directly that the data comes from a single example, rather than points to be iterated.

Unless I'm missing something, it would be much simpler to redefine all streams to operate on subsets of data which are shaped to be valid input to the learner.  (Single examples will then be 1-by-whatever-by-whatever.)

This change will simplify a bit of existing code, but also make it simpler to chain stream operators, such as `buffer_stream`.

@ejhumphrey any thoughts?  Would this terribly break your setup?  Does it introduce some terrible conceptual problem that i'm not seeing?

**Comment by ejhumphrey** [comment]
So. Allow me to say things that may or may not be dumb:

Streams could, as you say, produce stuff of the form:

- `x[i]`
- `x[i], y[i]`

But lets think about how this gets used. What we (generally) want to say is:

- `clf.partial_fit(x[i])`
- `clf.partial_fit(x[i], y[i])`

So, saying it more generally:

- `clf.partial_fit(*stream.next())`
- `clf.partial_fit(*stream.next())`

(hang tight, I'm going somewhere with this)

The next logical step in my head (and what would be great for me), then, is if we could actually break this out to (*arg, **kwarg) generators, which is what I've spent the last several months hacking together in a variety of unsatisfactory ways. Where this gets kinda hairy is if ...

1. You want to buffer streams into chunks / blocks / batches, or 
2. Your streams have different sample rates.

And I think (2) is an issue in the "operate on subsets of data" idea. Another problem, which would immediately break everything for me, is when you start mux'ing streams of streams.

Now that I've got that out, I also realize I'm not perfectly clear on what you're proposing.

**Comment by bmcfee** [comment]
> The next logical step in my head (and what would be great for me), then, is if we could actually break this out to (arg, *kwarg) generators, which is what I've spent the last several months hacking together in a variety of unsatisfactory ways.

can you elaborate on that?  Maybe an example?


> Now that I've got that out, I also realize I'm not perfectly clear on what you're proposing.

I guess what I most want is to retrofit things so that [buffer_stream](https://github.com/bmcfee/pescador/blob/master/pescador.py#L118) to act transparently.  The reason for this is that it would make it very simple and clean to sequester data batching (which is relatively costly) in the ZMQ worker process rather than doing it inside the StreamLearner.

The more I think about it, the more sense it makes to require streams to kick out `dict(X=...,Y=...)`s rather than tuples, so that we can keep a handle on variable names without having to do a bunch of contortions to determine whether it's supervised or not.  ('Y' can be optional in that case.)

Basically, I just want to enforce more structure on the data stream primitive to simplify the blocks upstream.  Does that make sense?

**Comment by ejhumphrey** [comment]
*Example*
https://github.com/ejhumphrey/biggie/blob/master/util.py#L85

Assume our streams yield ``entities``, which are basically structs with batteries. For the sake of discussion, we can think of buffering like populating a table, since all samples should have the same fields. When buffering structured data though (adding rows), you at some point need to define what your headers are. The function above does so by inspecting the data, with fingers crossed you haven't done something dumb. I wouldn't call this a good answer ... just an answer.

> Basically, I just want to enforce more structure on the data stream primitive to simplify the blocks upstream. Does that make sense?

Indeedy. I suppose we want / need a stream schema... so far it's been implicit rather than explicit. Things will get really fun if the streams / filters modify the structure of the data, e.g. ``(X, Y)`` in, ``(Z,)`` out. This is what the entity approach was trying to handle, at least then it's ``entity{X, Y}`` in, ``entity{Z}`` out, but this was still implicit to the functions in the chain.

**Comment by bmcfee** [comment]
I think we might be talking about two slightly different things.

1. What type of data (ie fields) does the generator produce, and how do those propagate through pescador?  Unless I'm mistaken, that's what your biggie link is doing, and I'm all for that type of pattern.  Data should be key-valued, ie, stored in a dict.

2. What are the semantics of the data in terms of samples and batches?  This is really what I'm trying to get at.  Right now, we assume that the input streams kick out individual samples, which are eventually bundled into batches.  This makes a certain kind of sense in the context of sampling IID from bags, but I'm not sure it leads to the most elegant implementation.

Instead, I'm thinking that we should change the requirements such that every item generated is a collection of samples.  (Note that that collection can be of size 1, reducing to the previously implemented case.)

The reason to do this is that it's more consistent in terms of the inputs and outputs of streams, and lets streams feed directly into `partial_fit()` without buffering.  Moreover, it will reduce communication overhead when sending iterates across the zmq channel, so learning should be faster.

Another benefit here is that the number of samples in the iterate is always explicit, and `X[i]` always selects the `i`th sample from an iterate.  If the sample index is implicit/suppressed, indexing data becomes ambiguous when it's multidimensional.

**Comment by bmcfee** [comment]
Jotting down some thoughts on the next iteration (yuk yuk).

1. The data generator primitive will be required to emit `batches`.  A `batch` is a `dict(X=..., [Y=...])`, where `len(X) == n_samples [== len(Y)]`.

2. The `Streamer` object will remain largely unchanged, excepting that it will now generate batches instead of samples

3. `mux()` will be lifted up to a `StreamMux` class, which contains an array of `Streamer`s.  `StreamMux` will itself implement an iterable interface, turning `n` streams into one.

4. `StreamLearner` will be simplified.  Its only novel method will be `iter_fit` which accepts a generator, and calls `partial_fit` on each iterate.  This way, the `Streamer` and `StreamMux` steps are essentially transparent.

5. `buffer_data` and `buffer_stream` will be removed and replaced with a new generator `buffer_batch(iterator, batch_size)`.  All this will do is collect iterates from `iterator` and group them into  larger batches.  `buffer_data` will of course also generate batches.

6. `mp_mux` and `zmq_mux` will be simplified and replaced by `mp_stream` and `zmq_stream`.  Instead of taking a seed pool and constructing a mux in the worker process, they will instead take a stream object that executes within the worker.  The advantages here are A) separating mux from process logic, and B) making it easy to stick a buffer in the worker process to reduce IPC.  These stream functions will, of course, generate batches.

**Comment by bmcfee** [comment]
First cut: [done!](http://nbviewer.ipython.org/github/bmcfee/pescador/blob/master/examples/Pescador%20demo.ipynb)

A few caveats:

- sparse matrix support is gone (for now)
- multiprocessing support is gone (for now)
- something strange is happening when passing labeled batches of `n=1` through zmq.  I really don't understand it, and feel like it might be a bug.
- we don't need a mux object.  mux is a generator, and generators can be safely bundled in Streamers, so this is the right way to push it over the zmq socket.
